File: src/main/java/commands/slashadapters/adapters/HelpAdapter.java
package commands.slashadapters.adapters;

import commands.Category;
import commands.runnables.informationcategory.HelpCommand;
import commands.slashadapters.Slash;
import commands.slashadapters.SlashAdapter;
import commands.slashadapters.SlashMeta;
import constants.Language;
import core.TextManager;
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;

@Slash(command = HelpCommand.class)
public class HelpAdapter extends SlashAdapter {

    public CommandData addOptions(CommandData commandData) {
        OptionData optionCategory = new OptionData(OptionType.STRING, "category", "Browse a command category", false);
        for (Category category : Category.values()) {
            optionCategory = optionCategory
                    .addChoice(TextManager.getString(Language.EN.getLocale(), TextManager.COMMANDS, category.getId()), category.getId());
        }

        return commandData
                .addOptions(optionCategory)
                .addOption(OptionType.STRING, "command", "Look up a specific command", false);
    }

    @Override
    public SlashMeta process(SlashCommandEvent event) {
        OptionMapping commandOption = event.getOption("command");
        OptionMapping categoryOption = event.getOption("category");

        String args = "";
        if (commandOption != null) {
            args = commandOption.getAsString();
        } else if (categoryOption != null) {
            args = "cat:" + categoryOption.getAsString();
        }

        return new SlashMeta(HelpCommand.class, args);
    }

}

File: src/main/java/commands/slashadapters/adapters/HelpAdapter.kt
package commands.slashadapters.adapters

import commands.Category
import commands.CommandContainer
import commands.CommandManager
import commands.runnables.informationcategory.HelpCommand
import commands.slashadapters.Slash
import commands.slashadapters.SlashAdapter
import commands.slashadapters.SlashMeta
import core.TextManager
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.Command
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData

@Slash(command = HelpCommand::class)
class HelpAdapter : SlashAdapter() {

    public override fun addOptions(commandData: SlashCommandData): SlashCommandData {
        var optionCategory = generateOptionData(OptionType.STRING, "category", "help_category", false)
        for (category in Category.values()) {
            optionCategory = optionCategory
                .addChoices(generateChoice(TextManager.COMMANDS, category.id, category.id))
        }
        return commandData.addOptions(
            optionCategory,
            generateOptionData(OptionType.STRING, "command", "help_command", false, true)
        )
    }

    override fun process(event: SlashCommandInteractionEvent): SlashMeta {
        val commandOption = event.getOption("command")
        val categoryOption = event.getOption("category")
        var args = ""
        if (commandOption != null) {
            args = commandOption.asString
        } else if (categoryOption != null) {
            args = "cat:" + categoryOption.asString
        }
        return SlashMeta(HelpCommand::class.java, args)
    }

    override fun retrieveChoices(event: CommandAutoCompleteInteractionEvent): List<Command.Choice> {
        val userText = event.focusedOption.value
        val choiceList = ArrayList<Command.Choice>()
        for (clazz in CommandContainer.getFullCommandList()) {
            val commandProperties = commands.Command.getCommandProperties(clazz)
            val commandTrigger = commandProperties.trigger
            val triggers = mutableListOf(commandTrigger)
            if ((!commandProperties.nsfw || event.channel!!.asTextChannel().isNSFW) &&
                CommandManager.commandIsTurnedOnEffectively(clazz, event.member, event.channel!!.asTextChannel())
            ) {
                triggers.addAll(commandProperties.aliases)
                if (triggers.any { it.lowercase().contains(userText.lowercase()) }) {
                    choiceList += Command.Choice(commandTrigger, commandTrigger)
                }
            }
        }

        return choiceList.toList()
            .sortedBy { it.name }
    }

}